source/grammar/qasm3Parser.g4

```diff
@@ -0,0 +1,361 @@
+    ;
+
+quantumMeasurement
+    : MEASURE indexIdentifier
+    ;
+
+quantumMeasurementAssignment
+    : quantumMeasurement (ARROW indexIdentifier)?
+    | indexIdentifier EQUALS quantumMeasurement
+    ;
+
+quantumBarrier
+    : BARRIER indexIdentifierList?
+    ;
+
+quantumGateModifier
+    : (INV | powModifier | ctrlModifier) AT
+    ;
+
+powModifier
+    : POW LPAREN expression RPAREN
+    ;
+
+ctrlModifier
+    : (CTRL | NEGCTRL) ( LPAREN expression RPAREN )?
+    ;
+
+quantumGateCall
+    : quantumGateModifier* quantumGateName ( LPAREN expressionList RPAREN )? indexIdentifierList
+    ;
+
+/*** Classical Instructions ***/
+
+unaryOperator: TILDE | EXCLAMATION_POINT | MINUS;
+
+expressionStatement
+    : expression SEMICOLON
+    ;
+
+expression
+    // include terminator/unary as base cases to simplify parsing
+    : expressionTerminator
+    | unaryExpression
+    // expression hierarchy
+    | logicalAndExpression
+    | expression DOUBLE_PIPE logicalAndExpression
+    ;
+
+/**  Expression hierarchy for non-terminators. Adapted from ANTLR4 C
+  *  grammar: https://github.com/antlr/grammars-v4/blob/master/c/C.g4
+  * Order (first to last evaluation):
+    Terminator (including Parens),
+    Unary Op,
+    Multiplicative
+    Additive
+    Bit Shift
+    Comparison
+    Equality
+    Bit And
+    Exlusive Or (xOr)
+    Bit Or
+    Logical And
+    Logical Or
+**/
+
+logicalAndExpression
+    : bitOrExpression
+    | logicalAndExpression DOUBLE_AMPERSAND bitOrExpression
+    ;
+
+bitOrExpression
+    : xOrExpression
+    | bitOrExpression PIPE xOrExpression
+    ;
+
+xOrExpression
+    : bitAndExpression
+    | xOrExpression CARET bitAndExpression
+    ;
+
+bitAndExpression
+    : equalityExpression
+    | bitAndExpression AMPERSAND equalityExpression
+    ;
+
+equalityExpression
+    : comparisonExpression
+    | equalityExpression EqualityOperator comparisonExpression
+    ;
+
+comparisonExpression
+    : bitShiftExpression
+    | comparisonExpression ComparisonOperator bitShiftExpression
+    ;
+
+bitShiftExpression
+    : additiveExpression
+    | bitShiftExpression BitshiftOperator additiveExpression
+    ;
+
+additiveExpression
+    : multiplicativeExpression
+    | additiveExpression (PLUS | MINUS) multiplicativeExpression
+    ;
+
+multiplicativeExpression
+    // base case either terminator or unary
+    : powerExpression
+    | unaryExpression
+    | multiplicativeExpression (ASTERISK | SLASH | PERCENT) ( powerExpression | unaryExpression )
+    ;
+
+unaryExpression
+    : unaryOperator powerExpression
+    ;
+
+powerExpression
+    : expressionTerminator
+    | expressionTerminator DOUBLE_ASTERISK powerExpression
+    ;
+
+expressionTerminator
+    : Constant
+    | Integer
+    | RealNumber
+    | ImagNumber
+    | BooleanLiteral
+    | Identifier
+    | StringLiteral
+    | builtInCall
+    | externOrSubroutineCall
+    | timingIdentifier
+    | LPAREN expression RPAREN
+    | expressionTerminator LBRACKET expression RBRACKET
```

Comment:
I suggest we should refine this rule a bit. Currently, there's an overlap between this line and `IndexIdentifier`, thus, sometimes the same string, say `a[10]`, can be parsed to either this or `IndexIdentifier`, depending on the location in the program.

Also, this line excludes expressions like `q[1:3]`, `q[1, 2, 3]` (since rangeDefinition is not an expresssion) but `IndexIdentifier` includes them. I am not sure if it's intended. It seems to me that those are useful to be in `expression`.

Looks like the only `expressionTerminator` can show up in this line are `Identifier` and `externOrSubroutineCall`. If we can make these 2 rules explict and potentially include `IndexIdentifier` into `expression` (depending on if we want things `q[1:3]` to be `expression`), it will make the grammar more clear.

Reply:
This comment chain was obsoleted by changes to how indexing is handled during the arrays PR (#283) - `indexExpression` became a complete part of the expression tree (with very tight priority) because we needed to allow the grammatical possibility of multiple indexers on the same identifier, and a specific rule was added for indexed identifiers in l-value positions.